### DISCUSSION TITLE
Recommendations for catching-up after extended downtime?

### DISCUSSION THREAD

**Discussion by kilt88** [zrepl/zrepl#625] [Q&A] [answered]
OS: omnios-r151040 (Illumos)
client: zrepl version=v0.5.0-dirty go=go1.17.5 GOOS=illumos GOARCH=amd64 Compiler=gc
server: zrepl version=v0.5.0-dirty go=go1.17.5 GOOS=illumos GOARCH=amd64 Compiler=gc

Here's a "fun" situation that maybe someone can help me with.

I run a push (tcp) job to a local sink (tcp) from one pool to another located in the same server.  Sockets don't work for me -- this is Illumos.

The target pool runs zfs native encryption.  After a reboot in April, I neglected to load the zfs encryption keys, and didn't notice this until July, that zrepl had been complaining about its inability to perform any replication.

I replicate a handful of filesystems, which until last week had been running 15-minute snapshots.  So you can imagine how back-logged the replication is.  When I restart zrepl, it can take 2 to 3 hours to to work through all the planning steps, and replication moves along extremely slowly.  I've since set snapshots to hourly.

I've played with concurrency steps and size_estimates, and I've brought the server to its knees (50 steps at once, for example).

I'm looking for an approach to get completely caught up, without the apparent months it will take to complete.  I could just destroy all snapshots and bookmarks and start over, but I'd like to understand better how to deal with this sort of situation.

Here is a sample listing of the number of snapshots in each filesystem that still need to be replicated before any purging can happen:

```
tank/fsA    10636
tank/fsB    10646
tank/images    10645
tank/public     7628
tank/fsC    10646
tank/tools     4524
tank/zones/plex     7796
```

zrepl.conf:
```
global:
  logging:
    - type: "stdout"
      level: "warn"
      format: "human"
  control:
    sockpath: /var/run/zrepl/control
  serve:
    stdinserver:
      sockdir: /var/run/zrepl/stdinserver
  monitoring:
    - type: prometheus
      listen: '192.168.xxx.xxx:9811'
      listen_freebind: false  # optional, default false

jobs:

- name: local_sink
  type: sink
  serve:
    type: tcp
    listen: "localhost:8888"
    clients: {
      "127.0.0.1" : "stor1"
      }
  recv:
    properties:
      inherit:
        - keylocation
        - keyformat
        - encryption
      override:
        canmount: noauto
        mountpoint: none
  root_fs: "backup/zrepl"

- name: backup_local
  type: push
  connect:
    type: tcp
    address: "localhost:8888"
  filesystems: {
    "rpool": true,
    "rpool/ROOT": true,
    "rpool/ROOT/r151038": true,
    "rpool/ROOT/r151040": true,
    "rpool/home": true,
    "tank": true,
    "tank/fsA": true,
    "tank/fsB": true,
    "tank/images": true,
    "tank/public": true,
    "tank/fsC": true,
    "tank/tools": true,
    "tank/zones": true,
    "tank/zones/plex": true
  }
  replication:
    concurrency:
      steps: 6
      size_estimates: 6
  snapshotting:
    type: periodic
    prefix: zrepl_
    interval: 1h
  pruning:
    keep_sender:
    - type: not_replicated
    - type: last_n
      count: 4
    - type: regex
      negate: true
      regex: "^zrepl.*"
    keep_receiver:
    - type: regex
      regex: ".*"

- name: remote_source
  type: source
  serve:
    type: tcp
    listen: "192.168.xxx.xxx:8889"
    clients: {
      "192.168.xxx.xxx" : "backupzrepl"
    }
  filesystems: {
    "backup/zrepl/stor1/rpool<": true,
    "backup/zrepl/stor1/tank<": true
  }
  send:
    encrypted: true
  snapshotting:
    type: manual
```

**Comment by problame** [comment, marked as answer]
Sorry for the late response, I guess it's irrelevant to you now but maybe someone will benefit from it in the future.

> I'm looking for an approach to get completely caught up, without the apparent months it will take to complete. I could just destroy all snapshots and bookmarks and start over, but I'd like to understand better how to deal with this sort of situation.


As long as you leave the `protection` knobs ([doc](https://zrepl.github.io/configuration/replication.html#protection-option)) at their default value, you can just manually destroy the snapshots on the sender that haven't been replicated yet.
If you get a `dataset is busy` on one of them, it's likely that it's the snapshot that is currently being replicated, and hence zrepl put a `zfs hold` on it. Lists with `zfs holds $snapshot`. Do not remove that hold. 
The concept of zrepl's holds and bookmarks is explained here:  https://zrepl.github.io/configuration/overview.html#zfs-abstractions-managed-by-zrepl

**Reply by kilt88** [reply]
I ended up just letting it run for 2 months, and it finally caught up.  :)

But next time -- if there is a next time -- I'll be sure to use your suggestions.

Thank you!
